plugins-root/check_icmp.c

```diff
@@ -1160,31 +1318,58 @@ add_target_ip(char *arg, struct in_addr *in)
 static int
 add_target(char *arg)
 {
-	int i;
-	struct hostent *he;
-	struct in_addr *in, ip;
+	int error, result;
+	struct sockaddr_storage ip;
+	struct addrinfo hints, *res, *p;
+	struct sockaddr_in *sin;
+	struct sockaddr_in6 *sin6;
+
+	switch (address_family) {
+	case -1:
+		// -4 and -6 are not specified on cmdline
+		address_family = AF_INET;
+		sin = (struct sockaddr_in *)&ip;
+		result = inet_pton(address_family, arg, &sin->sin_addr);
+#ifdef USE_IPV6
+		if( result != 1 ){
+			address_family = AF_INET6;
```

Comment:
address_family should be a local copy here. Otherwise address resolving might fail later. Consider this example:
`./check_icmp -H localhost -4`
address_family is -1 when entering add_target() but will be set to AF_INET6 during testing with inet_pton(). Then later getaddrinfo() uses AF_INET6 to resolve the name but fails because its AF_INET.

Reply:
i really think we should do something like this here:
```
result = inet_pton(AF_INET, arg, &sin->sin_addr);
if(result == 0)
    address_family = AF_INET;
```